### PR TITLE
Feat: test dep-trees generated by plugins as graphs

### PR DESCRIPTION
**Pull request by michael-go**
Use the new [@snyk/dep-graph](https://www.npmjs.com/package/@snyk/dep-graph) to convert the dep-tree generated by various plugins to a compact graph format, send it to the new `api/v1/test-dep-graph` endpoint, and handle the new response.

This PR doesn't change the behavior of testing `npm`/`yarn`, as it still happens in a different code path, and better be done in a separate PR to make this PR simpler.

This should resolve most issues related to large dependecy-trees that failed to scan (usually with with 413, 422, and 504 errors), such as: https://github.com/snyk/snyk/issues/159, https://github.com/snyk/snyk/issues/156

### CODE REVIEW

File: src/lib/plugins/rubygems/gemfile-lock-to-dependencies.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
+const gemfile = require('@snyk/gemfile');
+
+module.exports = gemfileLockToDependencies;
+
+const detectCycles = (dep, chain) => {
+  if (chain.indexOf(dep) >= 0) {
+    const error = Error('Cyclic dependency detected in lockfile', {dep, chain});
```
Comment (darmalovan) [inline src/lib/plugins/rubygems/gemfile-lock-to-dependencies.js]: Does the `{dep, chain}` is a thing as a second parameter to the error? never saw it
Maybe something like `err.meta.dep`...

File: src/lib/snyk-test/legacy.ts [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+function convertTestDepGraphResultToLegacy(
+    res,
+    depGraph: depGraphLib.DepGraph,
+    packageManager: string,
+    severityThreshold?: string): LegacyVulnApiResult {
+
+  const result = res.result;
+
+  const upgradePathsMap = [];
```
Comment (darmalovan) [inline src/lib/snyk-test/legacy.ts]: If it's a map shouldn't is be `{}` ?

File: package.json [hunk 3]
```diff
@@ -42,6 +42,8 @@
   "author": "snyk.io",
   "license": "Apache-2.0",
   "dependencies": {
+    "@snyk/dep-graph": "1.0.1",
```
Comment (darscan) [inline package.json]: we should probably update this to `1.1.0`

File: src/lib/snyk-test/run-test.ts [hunk 4]
```diff
@@ -146,27 +157,28 @@ async function assembleLocalPayload(root, options, policyLocations) {
       pkg.docker = pkg.docker || {};
       pkg.docker.baseImage = options['base-image'];
     }
+
+    if (_.get(pkg, 'files.gemfileLock.contents')) {
+      const gemfileLockBase64 = pkg.files.gemfileLock.contents;
+      const gemfileLockContents = Buffer.from(gemfileLockBase64, 'base64').toString();
+      console.time('gemfileLockToDependencies');
+      pkg.dependencies = gemfileLockToDependencies(gemfileLockContents);
+      console.timeEnd('gemfileLockToDependencies');
```
Comment (darmalovan) [inline src/lib/snyk-test/run-test.ts]: We still need this here?

Comment (gjvis) [inline src/lib/snyk-test/run-test.ts]: `console` still included? suggest to use `debug`?

Comment (michael-go) [reply src/lib/snyk-test/run-test.ts]: ooh it was there just for debug. @darscan we can simply remove this, right?

File: test/acceptance/cli.acceptance.test.js [hunk 5]
```diff
@@ -395,29 +649,27 @@ function (t) {
...
-        org: null,
-        packageManager: 'nuget',
-        path: 'nuget-app-2.1',
-        showVulnPaths: true,
-      },], 'calls nuget plugin');
-  });
-});
+  await cli.test('nuget-app-2.1');
 
+  var req = server.popRequest();
```
Comment (darmalovan) [inline test/acceptance/cli.acceptance.test.js]: constify?

File: test/acceptance/cli.acceptance.test.js [hunk 6]
```diff
@@ -1,7 +1,13 @@
-var test = require('tap').test;
+var tap = require('tap');
```
Comment (darmalovan) [inline test/acceptance/cli.acceptance.test.js]: constify? If you are here :)

File: test/acceptance/cli.acceptance.test.js [hunk 7]
```diff
@@ -305,52 +557,55 @@ test('`test npm-package-policy` returns correct meta', function (t) {
...
+      var req = server.popRequest();
+      t.equal(req.method, 'POST', 'makes POST request');
+      t.match(req.url, '/test-dep-graph', 'posts to correct url');
+
+      const depGraph = req.body.depGraph;
+      t.equal(depGraph.pkgManager.name, 'rubygems');
+      t.same(
+        depGraph.pkgs.map((p) => p.id).sort(),
+        ['ruby-app@', 'json@2.0.2', 'lynx@0.4.0'].sort(),
+        'depGraph looks fine');
```
Comment (gjvis) [inline test/acceptance/cli.acceptance.test.js]: what has this got to do with auto-detection of Gemfiles?

File: test/acceptance/cli.acceptance.test.js [hunk 8]
```diff
@@ -429,59 +681,54 @@ function (t) {
...
-    'specifies name');
-  });
+  await cli.test('monorepo', {file: 'sub-ruby-app/Gemfile'});
+
+  var req = server.popRequest();
+  t.equal(req.method, 'POST', 'makes POST request');
+  t.match(req.url, '/test-dep-graph', 'posts to correct url');
+  t.equal(req.body.depGraph.pkgManager.name, 'rubygems');
+  t.equal(req.body.targetFile, path.join('sub-ruby-app', 'Gemfile'),
+    'specifies target');
```
Comment (gjvis) [inline test/acceptance/cli.acceptance.test.js]: if this test is the overall "test a Gemfile in a monorepo" then perhaps depGraph based assertions (similar to the ones in the auto-detection test above) should be included here?

File: package.json [hunk 9]
```diff
@@ -27,7 +27,7 @@
     "eslint": "eslint -c .eslintrc src",
     "tslint": "tslint --project tsconfig.json --format stylish --exclude **/src/**/*.js",
     "prepare": "npm run build",
-    "tap": "tap test/*.test.* -Rspec --timeout=180 --node-path ts-node --test-file-pattern '/\\.[tj]s$/'",
+    "tap": "tap test/*.test.* -Rspec --timeout=300 --node-path ts-node --test-file-pattern '/\\.[tj]s$/'",
```
Comment (gjvis) [inline package.json]: this worries me. What fails with a `180` timeout?

File: src/lib/snyk-test/run-test.ts (hunk 4 above)
Comment (darscan) [reply src/lib/snyk-test/run-test.ts]: yup

File: src/lib/plugins/rubygems/gemfile-lock-to-dependencies.js (hunk 1 above)
Comment (darscan) [reply src/lib/plugins/rubygems/gemfile-lock-to-dependencies.js]: correct, current code is nonsense :)

File: src/lib/snyk-test/run-test.ts (hunk 4 above)
Comment (darscan) [reply src/lib/snyk-test/run-test.ts]: nope

File: src/lib/snyk-test/legacy.ts (hunk 2 above)
Comment (darscan) [reply src/lib/snyk-test/legacy.ts]: good catch

File: package.json (hunk 9 above)
Comment (darscan) [reply package.json]: worries me too.. what's changed?

File: src/lib/snyk-test/legacy.ts (hunk 2 above)
Comment (michael-go) [reply src/lib/snyk-test/legacy.ts]: 😱

File: src/lib/snyk-test/legacy.ts [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+  if (severityFilters.length) {
+    countText += ' ' + severityFilters.join(' or ') + ' severity';
+  }
+
+  return `${countText} vulnerable dependency ${pl('path', count)}`;
+}
+
+function pl(word, count) {
+  const ext = {
+    y: 'ies',
```
Comment (darmalovan) [inline src/lib/snyk-test/legacy.ts]: I don't if it's important but the `y` things is not being used

File: package.json (hunk 9 above)
Comment (michael-go) [reply package.json]: nothing in the code/tests we added actually - the tests were flaky due to timeouts for a while now. This is the tests that made me increase it: https://travis-ci.org/snyk/snyk/builds/460319719

File: package.json (hunk 3 above)
Comment (michael-go) [reply package.json]: done

File: src/lib/snyk-test/run-test.ts (hunk 4 above)
Comment (michael-go) [reply src/lib/snyk-test/run-test.ts]: done

File: src/lib/snyk-test/legacy.ts (hunk 2 above)
Comment (michael-go) [reply src/lib/snyk-test/legacy.ts]: fixed

File: src/lib/snyk-test/run-test.ts (hunk 4 above)
Comment (michael-go) [reply src/lib/snyk-test/run-test.ts]: fixed

File: src/lib/plugins/rubygems/gemfile-lock-to-dependencies.js (hunk 1 above)
Comment (michael-go) [reply src/lib/plugins/rubygems/gemfile-lock-to-dependencies.js]: fixed

File: test/acceptance/cli.acceptance.test.js (hunk 5 above)
Comment (michael-go) [reply test/acceptance/cli.acceptance.test.js]: I think we better do it in a separate chore PR just for easier diffing, as there are already lots-a changes

File: test/acceptance/cli.acceptance.test.js (hunk 6 above)
Comment (michael-go) [reply test/acceptance/cli.acceptance.test.js]: I think we better do it in a separate commit just for easier diffing, as there are already lots-a changes

File: test/acceptance/cli.acceptance.test.js (hunk 8 above)
Comment (michael-go) [reply test/acceptance/cli.acceptance.test.js]: done

File: test/acceptance/cli.acceptance.test.js (hunk 7 above)
Comment (michael-go) [reply test/acceptance/cli.acceptance.test.js]: not very much .. but basically it's a sanity test that we not only detected the `Gemfile` but also parsed the `Gemfile.lock`

**Pull request merged**